### PR TITLE
Upload

### PR DESCRIPTION
**Pull request by martindurant**

### CODE REVIEW

**Comment by martindurant** [comment]
@seibert @mmccarty , you might have thoughts about this

File: intake/container/__init__.py [hunk 1]
```diff
@@ -20,3 +20,40 @@
...
+def upload(data, path, **kwargs):
+    """Given a concrete data object, store it at given location return Source
+
+    Use this function to publicly share data which you have created in your
+    python session. Intake will try each of the container types, to see if
+    one of them can handle the input data, and write the data to the path
+    given, in the format most appropriate for the data type, e.g., parquet for
+    pandas or dask data-frames.
+
+    With the DataSource instance you get back, you can add this to a catalog,
```
Comment (martindurant) [inline intake/container/__init__.py]: Note that YAMLFileCatalog.add_source() will come later in this PR. I don't think it's meaningful for any other type of cat class (unless specifically developed for whatever service supplies the cat).

**Comment by danielballan** [comment]
Have you given much thought to how update semantics should work? Does `add` mean `upsert`?

File: intake/catalog/local.py [hunk 2]
```diff
@@ -578,6 +578,23 @@ def _load(self, reload=False):
                 text = text.replace('!template ', '')
             self.parse(text)
 
+    def add(self, **sources):
```
Comment (danielballan) [inline intake/catalog/local.py]: I would be inclined to do one `source` at a time here, akin to `set.add`.

Comment (martindurant) [reply intake/catalog/local.py]: I had wanted to give an explicit name; but I suppose a source may well have a name, so could be `add(self, source, name=None)` to provide the optional override.

**Comment by martindurant** [comment]
Right, `add`, as written, will clobber a source of the same name

**Pull request merged**